### ISSUE TITLE
Runtime std segfault with --static on windows

### ISSUE THREAD

**Issue by suirad** [ziglang/zig#1703]
Attempting to run any trivial program using `--static` on windows results in a segfault at runtime within the Std. 

Below are a couple examples:

---
### DirectAllocator.init()
```zig
const std = @import("std");
pub fn main() void{
    var da = std.heap.DirectAllocator.init();
    defer da.deinit();
}
```
Results:
![image](https://user-images.githubusercontent.com/8798829/48104157-d0241f00-e1f7-11e8-9d7d-7a9b768c00ab.png)

---

### getStderrStream()
```zig
const std = @import("std");
pub fn main() void{
    std.debug.warn("Boom\n");
}
```
Results:
![image](https://user-images.githubusercontent.com/8798829/48104229-285b2100-e1f8-11e8-9421-40c5486de72e.png)

**Comment by Sahnvour** [comment]
I've looked a bit into this and here's what I found.

The segmentation fault happens because we are trying to access memory that isn't readable, this is the `40096050h` constant on your screenshot. On your first example code, it gives something like this

![image](https://user-images.githubusercontent.com/1806407/54092191-755b2980-4389-11e9-9b6e-e98c8ab2cc09.png)

and if we drop the `--static` flag:

![image](https://user-images.githubusercontent.com/1806407/54092211-b2bfb700-4389-11e9-82ad-757f61de10f1.png)

That's the code to copy the function pointers to init the allocator. (I'm not sure why in the working case, the address is displayed as an offset from `RtlCaptureStackBackTrace`)

With `--static`, the address is clearly incorrect.

If we try to link with `link.exe` (command arguments from `--verbose-link`), there are many errors:
```
LINK : warning LNK4044: unrecognized option '/ERRORLIMIT:0'; ignored
LINK : warning LNK4216: Exported entry point WinMainCRTStartup
   Creating library static.lib and object static.exp
static.obj : warning LNK4210: .CRT section exists; there may be unhandled static initializers or terminators
builtin.lib(builtin.obj) : error LNK2017: 'ADDR32' relocation to '__unnamed_1' invalid without /LARGEADDRESSAWARE:NO
builtin.lib(builtin.obj) : error LNK2017: 'ADDR32' relocation to '__unnamed_2' invalid without /LARGEADDRESSAWARE:NO
builtin.lib(builtin.obj) : error LNK2017: 'ADDR32' relocation to '__unnamed_2' invalid without /LARGEADDRESSAWARE:NO
builtin.lib(builtin.obj) : error LNK2017: 'ADDR32' relocation to '__unnamed_3' invalid without /LARGEADDRESSAWARE:NO
builtin.lib(builtin.obj) : error LNK2017: 'ADDR32' relocation to '__unnamed_3' invalid without /LARGEADDRESSAWARE:NO
builtin.lib(builtin.obj) : error LNK2017: 'ADDR32' relocation to '__unnamed_3' invalid without /LARGEADDRESSAWARE:NO
builtin.lib(builtin.obj) : error LNK2017: 'ADDR32' relocation to '__unnamed_3' invalid without /LARGEADDRESSAWARE:NO
builtin.lib(builtin.obj) : error LNK2017: 'ADDR32' relocation to '__unnamed_3' invalid without /LARGEADDRESSAWARE:NO

... more errors ...

LINK : fatal error LNK1165: link failed because of fixup errors
```

So it seems there is a problem with the type of relocations of some symbols. Here's the result of `dumpbin /relocations builtin.lib` as an illustration: https://gist.github.com/Sahnvour/dbb6cb4f11abf292192fac52470129bf

Regarding the code in the example (`DirectAllocator.init`), we have:
```asm
std.heap.DirectAllocator.init:
.Lfunc_begin33:
	.cv_func_id 32
	.cv_file	9 "D:\\DEV\\Install\\zig\\lib\\zig\\std\\heap.zig"
	.cv_loc	32 9 50 0
.seh_proc std.heap.DirectAllocator.init
	pushq	%rbp
	.seh_pushreg 5
	movq	%rsp, %rbp
	.seh_setframe 5, 0
	.seh_endprologue
	movq	%rcx, %rax
.Ltmp454:
	.cv_loc	32 9 51 9
	movq	__unnamed_44, %rdx
	movq	%rdx, (%rcx)
	movq	__unnamed_44+8, %rdx
	movq	%rdx, 8(%rcx)
	movq	__unnamed_44+16, %rdx
	movq	%rdx, 16(%rcx)
	movq	__unnamed_44+24, %rdx
	movq	%rdx, 24(%rcx)
	popq	%rbp
	retq
```
```asm
__unnamed_44:
	.quad	std.heap.DirectAllocator.alloc
	.quad	std.heap.DirectAllocator.realloc
	.quad	std.heap.DirectAllocator.free
	.quad	
```
and in the dumpbin of the main object file:
```
 0000422B  ADDR32                     00000000        A1  __unnamed_44
 00004236  ADDR32                     00000008        A1  __unnamed_44
 00004242  ADDR32                     00000010        A1  __unnamed_44
 0000424E  ADDR32                     00000018        A1  __unnamed_44
```
This would be why the final executable has garbage addresses in it.

Unfortunately I lack the knowledge of linking to go further for now.

**Comment by andrewrk** [comment]
@Sahnvour thanks for looking into it. I think I realized the problem - Windows simply does not support non-Position Independent Code. This issue is closely related to #1828.

I ran into a similar issue with passing `-fPIC` to C source files on Windows, and received a compile error saying the option was not valid because all code is always PIC on Windows.

https://github.com/ziglang/zig/blob/fec4555476e38d2eeb1dfb02572404b243acd0b2/src/target.cpp#L1311-L1315

The `--static` CLI argument is overdue for some reworking.

**Comment by Sahnvour** [comment]
But how come non-PIC is generated without any error if that's invalid ?
Is this issue superseeded by #1828 ?

**Comment by andrewrk** [comment]
I'm not sure. I suppose LLVM is capable of doing it, so it does it, but then as you pointed out, the code can't really be linked meaningfully. Yes, I think that this issue will be solved by #1828.

**Comment by andrewrk** [comment]
`--static` is no longer an option that Zig offers. Instead, everything is as static as possible, and if you want something dynamic instead, use `-dynamic`.
